Update Theme to Use Bootstrap 3

thejuliekramer opened the pull request:
- [ ] Update Theme to Use Bootstrap 3 and CKAN 2.8
- [ ] Backwards compatible with Bootstrap 2 and CKAN 2.3
- [ ] Adds CircleCI

adborden commented on .circleci/config.yml:
```diff
@@ -0,0 +1,47 @@
+    working_directory: ~/ckanext-datagovtheme
+    machine:
+      image: circleci/classic:201708-01
+      # image: circleci/postgres:9.6-alpine
```
Can you remove any commented code?

adborden commented on .travis.yml:
```diff
@@ -0,0 +1,8 @@
+language: python
```
No need for a travis file.

adborden commented on README.md:
```diff
@@ -1,3 +1,31 @@
 # ckanext-datagovtheme
+
+[![CircleCI](https://circleci.com/gh/thejuliekramer/ckanext-datagovtheme.svg?style=svg)](https://circleci.com/gh/thejuliekramer/ckanext-datagovtheme)
```
This should be GSA's repo.

adborden commented on ckanext/datagovtheme/helpers.py:
```diff
@@ -526,3 +526,6 @@ def get_bureau_info(bureau_code):
 
     return bureau_info
+
+def ckan_version_28():
```
Sorry I didn't catch this earlier. Can we change this function to be more semantic? e.g. it's not really the version of CKAN that matters, it matters if we're using a CKAN version that supports bootstrap 2 vs 3. So `is_bootstrap2` might make more sense, and the default is to assume bootstrap 3.

For example, when CKAN 2.9 comes out, you shouldn't have to add a new function ckan_version_29().

adborden commented on bin/travis-run28.sh:
```diff
@@ -0,0 +1,6 @@
+
+echo "TESTING ckanext-datagovtheme 2.8"
+
+nosetests --ckan --with-pylons=test.ini ckanext/datagovtheme/tests/test_datagovtheme_28.py
```
This file seems strange to me. Should we always be running _all_ the tests?

thejuliekramer replied:
I split the tests because one is in an environment using 2.3 ckan and the other is in an environment using 2.8 ckan so only the tests for the applicable environment should be run

adborden replied:
Is it possible to stub the is_bootstrap2 method, render the template, and test the output? Then you would have several tests, one that stubs the method as version 2.3, another as 2.8, even one as 2.6 or 2.9. And all should pass no matter what the dependencies are installed,  including  CKAN.

I know it sounds a bit weird but we should be running the same tests against both versions... we want to test that our code does the right thing no matter what dependencies are installed.

thejuliekramer replied:
They are essentially the same test - but even the test setup is different due to the ckan version - I can combine the tests but will have to use a lot of if ckan_version statements. Is that ok?

adborden commented on test.ini:
```diff
@@ -0,0 +1,50 @@
+ckan.site_title = My Test CKAN Site
+ckan.site_description = A test site for testing my CKAN extension
+
+ckanext.geodatagov.dynamic_menu.url_default = http://www.data.gov/app/plugins/datagov-custom/wp_download_links.php
```
Should be https.

adborden commented:
Also for the datagovtheme_28.css file, this should be the default, since we'll be using bootstrap3 going forward. Let's rename this `datagovtheme.css` and the other one `datagovtheme_bootstrap2.css`.

Also, would you please remove any commented code from datagovtheme_28.css?

thejuliekramer commented on ckanext/datagovtheme/helpers.py:
```diff
@@ -4,7 +4,7 @@
 import StringIO
 
-from ckan.plugins.toolkit import config, request
+from pylons import config
```
It doesn't exist in 2.3

thejuliekramer commented on bin/travis-build.bash:
```diff
@@ -0,0 +1,103 @@
+pip install setuptools -U
+
+python setup.py develop
+cp ./ckan/public/base/css/main.css ./ckan/public/base/css/main.debug.css
```
It is required by CKAN 2.3 for some reason - the build fails without it if it doesn't exist. We can remove this once we move to 2.8 completely

adborden commented:
Then we should conditionally import with `check_ckan_version`
https://docs.ckan.org/en/2.8/extensions/plugins-toolkit.html#ckan.plugins.toolkit.ckan.plugins.toolkit.check_ckan_version

```python
if check_ckan_version(max_version='2.3'):
  from pylons import config
else:
  from ckan.plugins.toolkit import config
```

On Fri, Jan 10, 2020 at 8:05 AM Julie Kramer <notifications@github.com>
wrote:

> *@thejuliekramer* commented on this pull request.
> ------------------------------
>
> In ckanext/datagovtheme/helpers.py
> <https://github.com/GSA/ckanext-datagovtheme/pull/32#discussion_r365307685>
> :
>
> > @@ -4,7 +4,7 @@
>  import csv
>  import StringIO
>
> -from ckan.plugins.toolkit import config, request
> +from pylons import config
>
> It doesn't exist in 2.3
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/GSA/ckanext-datagovtheme/pull/32?email_source=notifications&email_token=AAD4OB32NAJGYGXXXWX5WELQ5CMEHA5CNFSM4KATJLVKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRLT23Y#discussion_r365307685>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD4OB6PKFFTZZYYDCMZ3H3Q5CMEHANCNFSM4KATJLVA>
> .
>


-- 
Aaron D Borden
Lead Engineer | IT Specialist
TTS | Data.gov <https://www.data.gov>

adborden commented:
Can you add a TODO comment then? "# TODO remove this after dropping 2.3
support. 2.3 fails without main.debug.css"

On Fri, Jan 10, 2020 at 9:13 AM Julie Kramer <notifications@github.com>
wrote:

> *@thejuliekramer* commented on this pull request.
> ------------------------------
>
> In bin/travis-build.bash
> <https://github.com/GSA/ckanext-datagovtheme/pull/32#discussion_r365339887>
> :
>
> > +	cd ckan
> +	git checkout 2.8
> +elif [ $CKANVERSION == '2.3' ]
> +then
> +	git clone https://github.com/GSA/ckan
> +	cd ckan
> +	git checkout datagov
> +fi
> +
> +echo "-----------------------------------------------------------------"
> +echo "Installing Python dependencies..."
> +
> +pip install setuptools -U
> +
> +python setup.py develop
> +cp ./ckan/public/base/css/main.css ./ckan/public/base/css/main.debug.css
>
> It is required by CKAN 2.3 for some reason - the build fails without it if
> it doesn't exist. We can remove this once we move to 2.8 completely
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/GSA/ckanext-datagovtheme/pull/32?email_source=notifications&email_token=AAD4OB4U3M7YS756MKXYM6TQ5CUCJA5CNFSM4KATJLVKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRL6HWA#discussion_r365339887>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD4OB46B3YGJKLZYSAQO4DQ5CUCJANCNFSM4KATJLVA>
> .
>


-- 
Aaron D Borden
Lead Engineer | IT Specialist
TTS | Data.gov <https://www.data.gov>

adborden commented:
Sure, let's pair on this for a bit.

On Fri, Jan 10, 2020 at 12:53 PM Julie Kramer <notifications@github.com>
wrote:

> *@thejuliekramer* commented on this pull request.
> ------------------------------
>
> In bin/travis-run28.sh
> <https://github.com/GSA/ckanext-datagovtheme/pull/32#discussion_r365425984>
> :
>
> > @@ -0,0 +1,6 @@
> +#!/bin/sh -e
> +
> +echo "TESTING ckanext-datagovtheme 2.8"
> +
> +nosetests --ckan --with-pylons=test.ini ckanext/datagovtheme/tests/test_datagovtheme_28.py
>
> They are essentially the same test - but even the test setup is different
> due to the ckan version - I can combine the tests but will have to use a
> lot of if ckan_version statements. Is that ok?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/GSA/ckanext-datagovtheme/pull/32?email_source=notifications&email_token=AAD4OBZ5Q6QEUZ67YRFSTV3Q5DN5LA5CNFSM4KATJLVKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRMZSBI#discussion_r365425984>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAD4OB2E2XGQAAROWAOJIOTQ5DN5LANCNFSM4KATJLVA>
> .
>


-- 
Aaron D Borden
Lead Engineer | IT Specialist
TTS | Data.gov <https://www.data.gov>

thejuliekramer commented:
@adborden I think I've addressed all comments other than running one set of tests for both 2.3 and 2.8 CKAN environments.  I tried to patch the method which returns whether CKAN is running bootstrap2 like so: https://github.com/thejuliekramer/ckanext-datagovtheme/pull/7/files but it is not working yet - maybe we can review together when you have time. Thanks.

amercader commented:
```python
if check_ckan_version(max_version='2.3'):
  from pylons import config
else:
  from ckan.plugins.toolkit import config
```

If you find yourselves doing this more than once then I recommend using [ckantoolkit](https://github.com/ckan/ckantoolkit) and just do:

```python
from ckantoolkit import config
```

That will work across all CKAN versions
